src/internal/connector/graph/service.go

```diff
@@ -319,3 +365,181 @@ func RunWithRetry(run func() error) error {
 
 	return support.ConnectorStackErrorTraceWrap(err, "maximum retries or unretryable")
 }
+
+// RetryHandler handles transient HTTP responses and retries the request given the retry options
+type RetryHandler struct {
+	// default options to use when evaluating the response
+	options RetryHandlerOptions
+}
+
+type RetryHandlerOptions struct {
+	// Callback to determine if the request should be retried
+	ShouldRetry func(delay time.Duration, executionCount int, request *http.Request, response *http.Response) bool
+	// The maximum number of times a request can be retried
+	MaxRetries int
+	// The delay in seconds between retries
+	DelaySeconds int
+}
```

Comment:
The indirection here (handler holding an options struct holding actual options) doesn't seem particularly helpful, except for the injection of all the extra interfaces associated with it. Personally, I think the code here could be greatly improved by cutting out all the abstracting and instead using values as plainly and straightforwardly as possible.

Start with a simple RetryHandler configuration, like so, and only add onto that as required.
```go
type RetryHandler struct {
  MaxRetries int
  MinimumBackoff time.Duration
}
```